packages/keystone/plugins/addressService.js

```diff
@@ -0,0 +1,114 @@
+const conf = require('@open-condo/config')
+const { ADDRESS_META_FIELD } = require('@open-condo/keystone/plugins/utils/addressMetaDefinition')
+const { composeResolveInputHook } = require('@open-condo/keystone/plugins/utils')
+const { plugin } = require('@open-condo/keystone/plugins/utils/typing')
+const { Text } = require('@keystonejs/fields')
+const {
+    createInstance: createAddressServiceClientInstance,
+    createTestInstance: createTestAddressServiceClientInstance,
+} = require('@open-condo/keystone/plugins/utils/address-service-client')
+const get = require('lodash/get')
+
+const readOnlyAccess = {
+    read: true,
+    create: false,
+    update: false,
+    delete: false,
+}
+
+const getSourceFieldName = (mainFieldName) => `${mainFieldName}Source`
+const getKeyFieldName = (mainFieldName) => `${mainFieldName}Key`
+const getMetaFieldName = (mainFieldName) => `${mainFieldName}Meta`
+
+const getFieldsToAdd = (mainField, fieldsHooks) => ({
+    [mainField]: {
+        schemaDoc: 'Normalized address',
+        type: Text,
+        isRequired: true,
+        hooks: {
+            ...(fieldsHooks[mainField] || {}),
+        },
+    },
+    [getSourceFieldName(mainField)]: {
+        type: Text,
+        schemaDoc: 'The origin of the address (some string which may looks like real address or some id)',
+        access: readOnlyAccess,
+        hooks: {
+            ...(fieldsHooks[getSourceFieldName(mainField)] || {}),
+        },
+    },
+    [getKeyFieldName(mainField)]: {
+        type: Text,
+        schemaDoc: 'The unique key of the address',
+        access: readOnlyAccess,
+        hooks: {
+            ...(fieldsHooks[getKeyFieldName(mainField)] || {}),
+        },
+    },
+    [getMetaFieldName(mainField)]: ADDRESS_META_FIELD,
+})
+
+/**
+ * Plugin to extend address field with address service
+ * @param {string?} addressFieldName
+ * @param {Object} fieldsHooks
+ */
+const addressService = (addressFieldName = 'address', fieldsHooks = {}) => plugin(({
+    fields = {},
+    hooks = {},
+    ...rest
+}, schemaObj) => {
+    if (!addressFieldName) {
+        throw new Error('The `addressFieldName` field is not set')
+    }
+
+    //
+    // Modify fields
+    //
+    const fieldsToAdd = getFieldsToAdd(addressFieldName, fieldsHooks)
+
+    // Check fields before adding new ones
+    Object.keys(fieldsToAdd).forEach((fieldNameToAdd) => {
+        if (fields[fieldNameToAdd]) {
+            throw new Error(`Plugin \`addressService\` can not add field \`${fieldNameToAdd}\` to \`${schemaObj.schemaName}\`, because this field already defined`)
+        }
+    })
+
+    //
+    // Modify hooks
+    //
+    const newResolveInput = async ({ resolvedData, operation, existingItem, originalInput, context }) => {
+        // We will call to address service in the following cases:
+        if (
+            // 1. In the case of new property/resident creation
```

Comment:
I think it should be just one case with an address field!

Then you call the address suggestion service search and save the result inside others fields.

1) createProperty({ address: `some string` }) 
2) call address suggestion service and get `addressKey`, `addressMeta`, `normalizedAddress`
3) save `some string` -> `addressSource`, `addressKey` -> `addressKey`, `addressMeta` -> `addressMeta`, `normalizedAddress` -> `address`

Reply:
The first case is the most frequent. The others (no address key and new address) will help us update the address key field manually.